prometheus/collector.go

```diff
@@ -29,24 +29,33 @@ type Collector interface {
 	// collected by this Collector to the provided channel and returns once
 	// the last descriptor has been sent. The sent descriptors fulfill the
 	// consistency and uniqueness requirements described in the Desc
-	// documentation. (It is valid if one and the same Collector sends
+	// documentation. It is valid if one and the same Collector sends
 	// duplicate descriptors. Those duplicates are simply ignored. However,
-	// two different Collectors must not send duplicate descriptors.) This
-	// method idempotently sends the same descriptors throughout the
-	// lifetime of the Collector. If a Collector encounters an error while
-	// executing this method, it must send an invalid descriptor (created
-	// with NewInvalidDesc) to signal the error to the registry.
+	// two different Collectors must not send duplicate descriptors. It is
+	// also valid to send no descriptor at all, which marks the Collector as
+	// unchecked, i.e. no checks will be performed at registration time, and
+	// the Collector may yield any Metric it sees fit in its Collect
+	// method. (Since the library will still ensure metric consistency,
+	// errors will thus only happen at scrape time. Thus, the Collector is
+	// still responsible to return metrics that are consistent with all the
+	// other metrics. Unchecked Collectors are useful if the exact metrics
+	// collected later cannot be predicted while consistency can still be
+	// ensured by the implementation.) This method idempotently sends the
+	// same descriptors throughout the lifetime of the Collector. If a
+	// Collector encounters an error while executing this method, it must
+	// send an invalid descriptor (created with NewInvalidDesc) to signal
+	// the error to the registry.
```

Comment:
The comment is quite long and hard to read and digest. Using paragraphs for the different topics (consistency, idempotence, etc.) would be helpful already. Cutting down on the prose and moving tips and tricks somewhere else would be good as well.

Reply:
I have introduced paragraphs and moved the explanation of the high-level concept to the package doc comment.